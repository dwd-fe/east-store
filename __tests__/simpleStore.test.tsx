import { createStore } from '../'
import React, { useEffect } from 'react'

import renderer, { act } from 'react-test-renderer'

describe('simpleStore', () => {
  jest.useFakeTimers()

  test('initialstate is atomic', () => {
    // atomic store
    const AtomicStore = createStore(0, {
      increase: () => count => count + 1,
      decrease: n => count => count - n
    })

    const Counter: React.FC = () => {
      const [count, action] = AtomicStore.useState()
      const handleDecrease = () => {
        expect(action.decrease(3)).toBe(void 0)
      }
      return (
        <div>
          <span>{count}</span>
          <button id="increase-btn" onClick={action.increase}>
            increase
          </button>
          <button id="decrease-btn" onClick={handleDecrease}>
            decrease
          </button>
        </div>
      )
    }
    const component = renderer.create(<Counter />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    for (let i = 0; i < 3; i++) {
      // increse 1
      act(component.root.findByProps({ id: 'increase-btn' }).props.onClick)
    }

    let increaseTree = component.toJSON()
    expect(increaseTree).toMatchSnapshot()
    // decrease 3
    act(() => {
      component.root.findByProps({ id: 'decrease-btn' }).props.onClick()
    })

    let decreaseTree = component.toJSON()
    expect(decreaseTree).toMatchSnapshot()
  })

  test('initialstate is complex', () => {
    const amy = {
      name: 'Amy',
      total: 130,
      score: { math: 60, english: 70 }
    }
    const buildStudentStore = (student: typeof amy) =>
      createStore(student, {
        modify: (subject: 'math' | 'english', score: number) => student => {
          student.score[subject] = score
          student.total = student.score.math + student.score.english
        }
      })

    const TestResult: React.FC<{
      studentStore: ReturnType<typeof buildStudentStore>
    }> = ({ studentStore }) => {
      const [{ name, total, score }, action] = studentStore.useState()
      useEffect(() => {
        setTimeout(() => {
          action.modify('math', 90)
        }, 500)
      }, [])
      return (
        <div>
          <span>name: {name}</span>
          <span>mathScore: {score.math}</span>
          <span>englishStore: {score.english}</span>
          <span>totalStore: {total}</span>
        </div>
      )
    }
    const amyStore = buildStudentStore(amy)
    const component = renderer.create(<TestResult studentStore={amyStore} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    expect(setTimeout).toHaveBeenCalledTimes(1)
    act(() => {
      jest.runAllTimers()
    })

    let afterTree = component.toJSON()
    expect(afterTree).toMatchSnapshot()
  })

  test('async action', async () => {
    const fetchRemoteTime = jest.fn(async () => {
      return await new Date(1573270146704)
    })

    const timer = createStore(new Date(1573270100000), {
      check: () => async () => {
        return await fetchRemoteTime()
      }
    })

    const Clock: React.FC = () => {
      const [date, action] = timer.useState()
      useEffect(() => {
        setTimeout(() => {
          expect(action.check()).toBe(void 0)
        }, 500)
      }, [])
      return <div>{date.toString()}</div>
    }

    const component = renderer.create(<Clock />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    await act(async () => {
      jest.runAllTimers()
      await fetchRemoteTime.mock.results[0].value
    })
    expect(fetchRemoteTime).toBeCalled()

    let afterTree = component.toJSON()
    expect(afterTree).toMatchSnapshot()
  })
})
