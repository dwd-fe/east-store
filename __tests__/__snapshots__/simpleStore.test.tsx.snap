// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simpleStore async action 1`] = `
<div>
  Sat Nov 09 2019 11:28:20 GMT+0800 (GMT+08:00)
</div>
`;

exports[`simpleStore async action 2`] = `
<div>
  Sat Nov 09 2019 11:29:06 GMT+0800 (GMT+08:00)
</div>
`;

exports[`simpleStore initialstate is atomic 1`] = `
<div>
  <span>
    0
  </span>
  <button
    id="increase-btn"
    onClick={[Function]}
  >
    increase
  </button>
  <button
    id="decrease-btn"
    onClick={[Function]}
  >
    decrease
  </button>
</div>
`;

exports[`simpleStore initialstate is atomic 2`] = `
<div>
  <span>
    3
  </span>
  <button
    id="increase-btn"
    onClick={[Function]}
  >
    increase
  </button>
  <button
    id="decrease-btn"
    onClick={[Function]}
  >
    decrease
  </button>
</div>
`;

exports[`simpleStore initialstate is atomic 3`] = `
<div>
  <span>
    0
  </span>
  <button
    id="increase-btn"
    onClick={[Function]}
  >
    increase
  </button>
  <button
    id="decrease-btn"
    onClick={[Function]}
  >
    decrease
  </button>
</div>
`;

exports[`simpleStore initialstate is complex 1`] = `
<div>
  <span>
    name: 
    Amy
  </span>
  <span>
    mathScore: 
    60
  </span>
  <span>
    englishStore: 
    70
  </span>
  <span>
    totalStore: 
    130
  </span>
</div>
`;

exports[`simpleStore initialstate is complex 2`] = `
<div>
  <span>
    name: 
    Amy
  </span>
  <span>
    mathScore: 
    90
  </span>
  <span>
    englishStore: 
    70
  </span>
  <span>
    totalStore: 
    160
  </span>
</div>
`;
